### PR TITLE
[frontend] Sign Up api

### LINKED ISSUE #94: [frontend] Calling API
## U know rule and so do i
✔️: done and merged
⌛: pr and waiting for review
❌: not finished

### (A) Discover & News TYK ✔️
- [x] /news/newsID
- [x] /discover?pageNumber={pageNumber}

### (B) ViewShop TYK ⌛
- [x] /sellerID/shop
- [x] /seller/manageProducts
- [x] /seller/manageProducts/new
- [x] /seller/manageProducts/productID

### (C) Order TYK ⌛
- [x] /buyer/order
- [x] /buyer/order/orderID
- [x] /seller/order/orderID
- [x] /seller/order

### (D) Product TYK ⌛
- [x] /product

### (E) Coupon UG ⌛
- [x] /admin/manageCoupon
- [x] /admin/manageCoupon/couponID
- [x] /admin/manageCoupon/new
- [x] /seller/manageCoupon
- [x] /seller/manageCoupon/new
- [x] /seller/manageCoupon/couponID
- [x] /sellerCoupon

### (F) Report TYK ✔️
- [x] /seller/report
- [x] /seller/report/reportID
- [x] /admin/report
- [x] /admin/report/reportID

### (G) Cart UG ❌
- [x] /buyer/cart 

### (H) Login & Signup UG ✔️
- [x] /login
- [x] /signup

### (I) Info Settings UG ✔️
- [x] /user/info
- [x] /seller/info

### (J) Admin ManageUser UG ⌛
- [x] /admin/manageUser 

### (K) Security UG ✔️
- [x] /user/security
- [x] /user/security/password
- [x] /user/security/manageCreditCard
- [x] /user/security/manageCreditCard/new

### (L) Search TYK ⌛
- [x] /search
- [-] /search?shop={keyword}
- [x] /search?keyword={keyword}

### PR DESCRIPTION
**Pull request by UGisBusy**
Resolves #94 

#### Page
- [x] [/signup](http://localhost:8080/signup)

#### APIs
- Customer signup

#### Function
- the account data have these constraints:
  - email: need to be an email
  - username: only english letter and number, less than 32 characters
  - name: no constraint
  - password: contain at least one upper case, one lower case, one number and one special character, 8 to 72 characters
- if some contraints are not fullfilled, it shouldnt send the data
- if all  contraints are fullfilled
  - will navigate to login and this means it success
  - will pop up warning modal that show the error from backend. (duplicated email or username)

#### Auth-related Error handling
- if password/username are wrong, pop up warning and re-enter again
- if 500 at auth or getting access token, naviagte to ``/login``

### Testing Approach (if unconventional)
```sh
docker compose up --build
```
- Go to login page
  - go to web [normally](http://localhost:8080/)
  - if ur in the main page, click the ``logout`` in ``navbar``
- Create an account
  - click ``Sign up``, than input things.
  - if success. will navigate to login page
- Login (Auth)
  -  click ``login`` in login page, than input things.

### Additional context
N/A

### CODE REVIEW

File: frontend/src/lib/Status.tsx
Comment (NOOBDY) [inline frontend/src/lib/Status.tsx:1]: should this be a `.tsx` file?

File: frontend/src/lib/Status.tsx [hunk 1]
```diff
@@ -1,7 +1,17 @@
+import { NavigateFunction } from 'react-router-dom';
+
+export const RouteOnNotOK = (responce: Response, navigate: NavigateFunction) => {
```
Comment (NOOBDY) [inline frontend/src/lib/Status.tsx]:
```suggestion
export const RouteOnNotOK = (response: Response, navigate: NavigateFunction) => {
```
I would suggest you install a spell checker tool

File: frontend/src/pages/user/signup.tsx [hunk 2]
```diff
@@ -28,57 +68,65 @@ const Signup = () => {
...
+                    <FormItem label='email'>
+                      <input type='email' {...register('email', { required: true })} />
+                    </FormItem>
+                    <FormItem label='username'>
+                      <input type='text' {...register('username', { required: true })} />
+                    </FormItem>
+                    <FormItem label='name'>
+                      <input type='text' {...register('name', { required: true })} />
+                    </FormItem>
+                    <FormItem label='Password'>
```
Comment (NOOBDY) [inline frontend/src/pages/user/signup.tsx]:
```suggestion
                    <FormItem label='password'>
```

File: frontend/src/pages/user/authorize.tsx [hunk 3]
```diff
@@ -1,37 +1,51 @@
...
+  const [show, setShow] = useState<boolean>(false);
+  const [warningText, setWarningText] = useState<string>('');
+  const navigate = useNavigate();
 
-  const [email, setEmail] = useState<string>('');
-  const [password, setPassword] = useState<string>('');
+  const { register, handleSubmit } = useForm<FormProps>();
 
   const authUrl = '/api/oauth/authorize';
+  const body = Object.fromEntries([...searchParams.entries()]);
```
Comment (NOOBDY) [inline frontend/src/pages/user/authorize.tsx]:
```suggestion
  const body = Object.fromEntries(searchParams.entries());
```

File: frontend/src/pages/user/callback.tsx [hunk 4]
```diff
@@ -33,21 +30,28 @@ const Callback = () => {
...
+    console.log('Missing code, state or verifier');
+    return <Navigate to='/login' replace={true} />;
+  }
+
+  refetch();
   if (isPending) {
     return <>Loading</>;
   }
 
   if (error) {
```
Comment (NOOBDY) [inline frontend/src/pages/user/callback.tsx]:
```suggestion
  const { isPending, isError, data, refetch } = useQuery({
    queryKey: ['token'],
    queryFn: getToken,
    enabled: false,
  });

  if (!code || !state || !verifier) {
    console.log('Missing code, state or verifier');
    return <Navigate to='/login' replace={true} />;
  }

  refetch();
  if (isPending) {
    return <>Loading</>;
  }

  if (isError) {
```
i think i effed up before

**Review by NOOBDY — CHANGES_REQUESTED** [review]

File: frontend/src/lib/Status.tsx
Comment (UGisBusy) [reply frontend/src/lib/Status.tsx:1]: Status.tsx will have this component after mergin main.
```tsx
export const CheckFetchStatus = ({ status }: FetchStatusProps) => {
  switch (status) {
    case 'pending':
      return <div>Loading...</div>;
    case 'error':
      return <div>Error...</div>;
  }
};
```
Or do u think putting them in the same file is bad?

**Review by KatoToyoki — COMMENTED** [review]
image

is it normal that when i sign up with a new account then login, encounter of the situation of 401?
edge, safari and firefox all have this

edit:
after reloading the page it will disappear but it always appears when first login

**Comment by UGisBusy** [comment]
I think one is when u first enter the website. 
And second is when u sign upped, navigate back to ``login``, than click the login button.
It's likely that this in intended imo.

**Review by KatoToyoki — APPROVED** [review]

**Review by NOOBDY — APPROVED** [review]

**Pull request merged**